### ISSUE TITLE
[windows dx] RenderTarget2D.SaveAsPng width & height parameters are unused

### ISSUE THREAD

**Issue by lofcz** [MonoGame/MonoGame#6642]
## Traced route

```csharp
        // just my code
        MyRendertarget.SaveAsPng(MyStream, 64, 64)

        // leads to
        public void SaveAsPng(Stream stream, int width, int height)
        {
            // Tracing this
            PlatformSaveAsPng(stream, width, height);
        }

        // width, height are not used
        private void PlatformSaveAsPng(Stream stream, int width, int height)
        {
            var pngWriter = new PngWriter();
            pngWriter.Write(this, stream);
        }
```

## Proposed solution
As `pngWriter.Write(this, stream);` works with Texture2D size, from my understanding it wouldn't be easy to patch it to work with w & h other than native to the passed Texture. Therefore my suggestion is to scale RenderTarget2D and pass it in proper size to `pngWriter.Write();` current implementation.

```csharp
        // Define a temp RenderTarget of correct size
        RenderTarget2D temp = new RenderTarget2D(GraphicsDevice, width, height);
        GraphicsDevice.SetRenderTarget(temp);  

        // Rectangle can either be defined here or pooled to prevent garbage generation     
        GeneralRectangle.X = 0;
        GeneralRectangle.Y = 0;
        GeneralRectangle.Width = width;
        GeneralRectangle.Height = height;

        // Drawback - extra batch call and user can't control arguments passed to sb.Begin 
        // also SpriteBatch "sb" would need to be passed
        sb.Begin();
        sb.Draw(temp, GeneralRectangle, Color.White);
        sb.End();
   
        // either to null or remember original RenderTarget
        GraphicsDevice.SetRenderTarget(null);

        // finally save result
        temp.SaveAsPng(stream, width, height);
```

Tested this patch to prove it working, but this is rather in-framework solution. There are several drawbacks mentioned, it may be more efficient to scale Texture directly in `pngWriter.Write` before passing it to Zlib.

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:
- WindowsDX

**Comment by lofcz** [comment]
@Jjagg Correct, I've forgot about this. Guess this is not a priority for you, I'll try to patch up a PR.

**Comment by Jjagg** [comment]
It is not. A PR would be greatly appreciated :) 
One thing to be considered is what kind of filter we should use when resizing. 
Another thought is that we could maybe use a pooled array to store resized data in.

**Comment by willmotil** [comment]
How do you get the pixel data out of the texture internally ? 
With get data ? If so does that work on all platforms ?

**Comment by rds1983** [comment]
btw, StbSharp project has stb_image_resize.h ported: https://github.com/rds1983/StbSharp/blob/master/StbSharp/StbImageResize.Generated.cs

Though maybe it's not rational to include file with 2k lines to MG to support feature that almost noone use(SaveAsPng with sizes).
But it's still a possible solution.

**Comment by Jjagg** [comment]
> How do you get the pixel data out of the texture internally ?
With get data ? If so does that work on all platforms ?

The back end graphics APIs expose a way to copy data from GPU memory to CPU RAM. Some lower end Android devices don't support this if I remember correctly.

> StbSharp project has stb_image_resize.h ported

Resizing is a pretty complex operation. If we implement this I think it'd be great to use StbImageResize.

**Comment by rds1983** [comment]
> Resizing is a pretty complex operation. If we implement this I think it'd be great to use StbImageResize.
Right now, MG uses quite outdated StbSharp framework.
Which is fixed by #6355.
If it gets merged, then adding StbImageResize(or any other StbSharp module) would mean just copying StbImageResize.cs and StbImageResize.Generated.cs to MG.

**Comment by Jjagg** [comment]
@rds1983 Oops, well that one has been left alone for far too long :sweat_smile:
